### ISSUE TITLE
macOS builds are not reproducible

### ISSUE THREAD

**Issue by samm-git** [smartmontools/smartmontools#213]
## Summary

Recently support for the reproducible builds was added to the smartmontools project, using (SOURCE_DATE_EPOCH)[https://reproducible-builds.org/docs/source-date-epoch/] environment variable to set the build date. At the moment macOS builds are not reproducible. This ticket is to track issues and progress on that issue

### Binary builds

Binary builds (`smartd` and `smartctl`) are reproducible by default if built on `osxcross`. However, if built locally linkder provides different binary on every build. This could be fixed by adding env `ZERO_AR_DATE` (see `src/ld/Options.cpp`).

### Building pkg

It was found that `xar` archiver used to build `pkg` package by default adds timestamp of the archive creation and atime/ctime/mtime information for every file. It is possible to disable *time data (as well as other noise) using `--distribution` switch, but the `creation-timestamp` is not changeable. I created [patch](https://github.com/tpoechtrager/xar/pull/3) to the osxcross fork. Using my patch and `--distribution` key checksums are matching.  

### Creating iso file

At the moment we are creating iso file (which is later converted to dmg) using `mkisofs` tool. It does not support reproducible builds. As a workaround we can migrate to the `xorriso` tool which respects `SOURCE_DATE_EPOCH` out of the box.

Command line to build iso provided below seems to provide reproducible builds with the same checksum:
```
xorriso -as mkisofs  -V 'smartmontools' -no-pad -r -o smartmontools-7.5.darwin/smartmontools-7.5.iso       -hfs-bless "smartmontools-7.5.darwin/dmg/" "smartmontools-7.5.darwin/dmg/"
```
### Converting iso to dmg

It seems that dmg utility works fine out of the box and does not add any build specific information. 

```
 dmg dmg smartmontools-7.5.darwin/smartmontools-7.5.iso test.dmg && md5sum test.dmg
```
always returns same checksum.

## Todo

- [x] set `ZERO_AR_DATE` to 1 if builds natively on macos and `SOURCE_DATE_EPOCH` is set
- [x] use xar with timesmap patch and apply `--distribution` key
- [x] migrate from `mkisofs` to `xorriso`
- [x] remove gzip timestamp when generating Payload
- [x] cpio always storing timestamp when creating payload, so we need to use `touch` to override
- [x] xorriso always storing SOURCE_DATE_EPOCH timestamp when creating payload
- [x] Validate that CI builds are reproducible

## Validation

- [x] Binaries (smartctl, smart) files are reproducible (clang, ld)
- [x] Payload in the pkg files is reproducible (cpio, gzip)
- [x] Pkg files are reproducible (xar)
- [x] ISO files are reproducible (xorriso) - **failed, we need to remove time from file info**
- [x] DMG files are reproducible (dmg)

**Comment by samm-git** [comment]
Done, conformed on CircleCI

**Comment by chrfranke** [comment]
Great! Thanks!

**Comment by chrfranke** [comment]
A minor issue regarding [r5549](https://www.smartmontools.org/changeset/5549) (5b98c6d):

The `export` statement is apparently not supported by BSD variants of `make`. These require `.export` instead. Tested with a local build of `bmake-20230723` from https://www.crufty.net/help/sjg/bmake.html and current `bmake` package from Debian 12.

Unfortunately GNU autoconf+automake do not provide a built in way to handle this incompatibility (IMO a missing feature). Could easily be handled by a custom check in `configure.ac` and some `@EXPORT@` instead of `export` in `Makefile.am`.

I guess this is not needed because GNU `make` should be available on macOS also. Thoughts?

Interestingly POSIX 2018 (https://pubs.opengroup.org/onlinepubs/9699919799/utilities/make.html) says:

"
FUTURE DIRECTIONS
Some implementations of _make_ include an _export_ directive to add specified _make_ variables to the environment. This may be considered for standardization in a future version.
"

**Comment by samm-git** [comment]
@chrfranke thank you for the heads up. I got your email about the same but was a bit busy. Yes, I also was thinking that this could be a gnu-specific extension. I don't think it would cause issues for us as it seems that macOS uses GNU make already:

```
 % /usr/bin/make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```
However, this file is likely a shim to the `/Library/Developer/CommandLineTools/usr/bin/make` file, which is installed as part of the XCode CommandLineTools set. So at least on a recent xcode `make` is `gnumake`. Probably on some previous versions it was not the case, because i also see `gnumake` binary in the same folder, which has the same checksum as `make`. 

So long story short - recent xcode is using gnu make, so using `export` statement should not cause any problems.

**Comment by chrfranke** [comment]
OK, thanks for checking this. Then we have no issue because this only affects reproducible builds for macOS. Any other build should also work with BSD make.
